### DISCUSSION TITLE
Any suggestion on how to use Apple 2 HGR?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1511] [Q&A]
I am trying to implement real Apple 2 HGR graphics in my Cross-Lib project.
I see that CC65 includes some specific cfg files but they seem to be meant for binaries that embed hgr pictures.

I would like to have a binary where I can draw in HGR mode. 
Do I need to create a new cfg with a "hole" in $2000-$3f00 and/or $4000-$4fff ? So that code avoids it?

If I use the default cfg I see a simple program (see below) meant to fill the screen to freeze once a certain row is filled.
I suppose it is overwriting itself.

```
#include <conio.h>
#include <peekpoke.h>

void _XL_INIT_GRAPHICS(void)
{
    
    // HIGH RES ON
    POKE(0xC057,0);

    // HGR PAGE 2
    // POKE(0xC055,0);

    // HGR PAGE 1
    POKE(0xC054,0);
 
    // TEXT OFF
    POKE(0xC050,0);
    
    // MIXED OFF
    POKE(0xC052,0);
    
    {
        uint8_t i;
        uint8_t j;
        
        for(i=0;i<192;++i)
        {
            for(j=0;j<40;++j)
            {
                POKE(HB1[i]+j,255);
            }
            cgetc();
        }
    }
}
```

What is wrong in my code?

**Comment by oliverschmidt** [comment]
> I see that CC65 includes some specific cfg files but they seem to be meant for binaries that embed hgr pictures.

No, they are rather meant for your use case ;-)

> Do I need to create a new cfg with a "hole" in $2000-$3f00 and/or $4000-$4fff ? So that code avoids it?

There's no DOS for the Apple II that allows to load several chunks of a program to different locations. So you have two options:

1. Have the DOS load your program to $4000 (or $6000) and make use of the memory between $0803 and $2000 (or $4000) in some other way. There's no need for a new .cfg file for that approach. Rather you can use the start-address feature to archive the desired result.

2. Have the DOS load your program to $0803 and make sure that $2000-$4000 (or $6000) doesn't contain anything your program needs to run. So your program actually needs to contain $2000 (or $4000) bytes to be loaded at $2000. Given that the DOS doesn't decompress programs during loading, it doesn't matter if the program contains $2000 zero's or an embedded HGR picture. Therefore users usually opt for the HGR picture. But independently from that choice, there's no need for a new .cfg file for that approach. Rather you can use a ...-hgr.cfg file to archive the desired result.

**Reply by Fabrizio-Caruso** [reply]
What does this practically mean for me?
Do I have to have a dummy .s file that starts with ".segment hgr" and then 8192 zeros?
If I do nothing else, I will have an empty area between $803 and $1fff but I don't mind but I want the load to start at $4000.
How do I do this? Do I have to compile with -s 16384?

**Reply by Fabrizio-Caruso** [reply]
What does "have the does load your program" practically mean? Isn't the DOS starting the binary at $0803 anyway?
Let us assume taht  I just want to use page 1 at $2000-$3FFF.
I will need more than just $0803-$1FFF, so I guess I will have to instruct the linker to put some of code at $4000-...

**Reply by Fabrizio-Caruso** [reply]
Maybe I got it.
This seems to work:
 
```
apple2_test:	
	$(CC65_PATH)$(MYCC65) $(INCLUDE_OPTS) -t apple2 \
	--config $(CFG_PATH)/cc65/apple2-hgr.cfg \
...
	--code-name LOWCODE \
	$(SOURCE_PATH)/end_screen.c \
	$(CROSS_LIB_PATH)/display/init_graphics/cc65/apple2/hgr.s \
	$(CROSS_LIB_PATH)/display/init_graphics/cc65/apple2/apple2_init_graphics.c \
	$(CROSS_LIB_PATH)/sound/generic/bit_bang_sounds.c \
	$(CROSS_LIB_PATH)/sleep/sleep_macros.c \
	--code-name CODE \
	$(FULL_FILES) \
	-o $(BUILD_PATH)/apple2.bin
```

**Reply by polluks** [reply]
You may use [Exomizer](https://bitbucket.org/magli143/exomizer/wiki/Home) to "pack the hole".

**Reply by Fabrizio-Caruso** [reply]
@oliverschmidt I am not sure I understand what I am doing... It sort of works but it also hangs for reasons I don't understand.
What is the purpose of LOWCODE? Just having some binary code in $0803-$1FFF?

**Comment by Fabrizio-Caruso** [comment]
My game will crash depending on which source files I put after `--code-name LOWCODE`.
My understanding is that, by default (no special options),  code can be in both LOWCODE and CODE and I can put anything in either memory area as long as there is no overflow. I assume that an overflow will produce a compilation error. I am assuming the code starts at $0803 even if no explicit code is in LOWCODE. I presume the crt is there and it will jump to the main function.

Maybe having only one code area is simpler but it does not work if I put nothing in LOWCODE.
Do I need to change the start address to get something working if nothing is in LOWCODE?

**Reply by Fabrizio-Caruso** [reply]
I fear that the apple2-hgr.cfg is not filling the LOWCODE segment. So it starts writing the HGR before $2000.
I would like to have some code in LOWCODE (and have LOWCODE filled with zeros). How do I do this?
I suppose I need to creat a memoy section and change the cfg but you said I don't have to do that. So I am not sure...

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt I guess the apple2_hgr.cfg can only work if I instruct somehow the linker to fill LOWCODE with zeros.
I am coding just in C.
If I use  `--code-name LOWCODE`  and   `--code-name CODE`  in my command linke in order to have some generated code in LOWCODE and some other generated code in CODE, how can I be sure that the linker is filling LOWCODE and not starting the HGR segment before $2000? 
I see start=$2000 for HGR in the cfg but I am not sure the linker uses it to fill the segment before (i.e., LOWCODE).

**Reply by polluks** [reply]
You may use LOWCODE if you don't want to waste the memory area $0803-$1FFF but it's not a requirement.
However I'm a Commodore guy.

**Reply by Fabrizio-Caruso** [reply]
@polluks, I don't mind using LOWCODE. The problem is that I don't understand how to use the default hgr config.

I am compiling with this command line:
```
apple2_hgr:	 
	$(CC65_PATH)$(MYCC65) $(INCLUDE_OPTS) -t apple2 \
	--config $(CFG_PATH)/cc65/apple2-hgr.cfg \
...
    --code-name LOWCODE \
	$(SOURCE_PATH)/end_screen.c \
	$(CROSS_LIB_PATH)/display/init_graphics/cc65/apple2/apple2_hgr_init_graphics.c \
    --code-name CODE \
	$(CROSS_LIB_PATH)/sound/generic/bit_bang_sounds.c \
	$(CROSS_LIB_PATH)/sleep/sleep_macros.c \
	$(CROSS_LIB_PATH)/display/init_graphics/cc65/apple2/hgr.s \
	$(FULL_FILES) \
	-o $(BUILD_PATH)/apple2.bin
```

Is this wrong? It looks some code is overwritten because my game only works if I reduce the code size.
How does the linker manage to fill the whole between LOWCODE and HGR ($2000-$3FFF).
I put some stuff in LOWCODE but how do I tell the linker to fill it till $1FFF?

Which options should I use to debug this?

**Comment by greg-king5** [comment]
The order of the source filenames in that command isn't important.  You can use the `--code-name` option only once.  Put
```
--code-name LOWCODE \
```
on a line.  Put _all_ files that you want in `CODE` above that line.  Put all files that you want in `LOWCODE` below that line.

ld65 will pad `LOWCODE` if it's too small.  But, ld65 won't warn you if it's too big!  ld65 simply will push the next segment up to a higher address -- not what you want!  You must check for that condition manually.  You must tell the command to make a map file.  Then, you must check the `Segment list` section of that map.  If the `HGR` line has a start address that's too high, then you must move some files from `LOWCODE` over to `CODE`.  On Windows, you can use the "calc" accessory (in Programmer view) to determine how many bytes you need to move.  You can look through the `Modules list` section of the map to see how big the `LOWCODE` and `CODE` segments are, in each input file.

You can rearrange the filenames in the Makefile recipe so that as much as possible is put into `LOWCODE` without going over the limit.

**Comment by greg-king5** [comment]
You can replace "hgr.s" with three lines:
```asm
.segment "HGR"

.res $2000, $00
```

**Reply by greg-king5** [reply]
Hmm ...
I remember that you want to avoid Assembly files as much as you can.  So, let's move the bitmap into "apple2_hgr_init_graphics.c".  Put these lines into that C file:
```c
#pragma bss-name(push, "HGR")
static uint8_t bitmap[0x2000];
#pragma bss-name(pop)
```
Remove "hgr.s".

**Reply by Fabrizio-Caruso** [reply]
Thanks! Your solution is much nicer than mine.
I don't mind having Assembly files in Cross-Lib implementation files. I prefer C because I am sharing some code across different CPU architectures. By the way, this is not easy sometimes because of CC65's lack of support for the `volatile` attribute.
I cannot use Assembly, nor target-specific and not even CC65-specific code game code as the the very purpose of Cross-Lib is to allow the build of dev-kit and hardware-agnostic games from the very same source code. I don't even allow explicit conditional compilation directives in the game code.
All the target specific code including Assembly in any sauce (some snippets of Assembly for several architectures are already part of Cross-Lib implementation).

**Reply by Fabrizio-Caruso** [reply]
@greg-king5 
If I try to follow your suggestion I get:
```
cross_lib/display/init_graphics/cc65/apple2/apple2_hgr_init_graphics.c(129): Warning: Variable 'bitmap
' is defined but never used
ld65: Warning: cross_lib/cfg//cc65/apple2-hgr.cfg(27): Segment 'CODE' start address is too low in 'MAI
N' by 8192 bytes
ld65: Warning: cross_lib/cfg//cc65/apple2-hgr.cfg(17): Segment 'CODE' overflows memory area 'MAIN' by
2274 bytes
ld65: Error: Cannot generate most of the files due to memory area overflows
```

What I have done is to have 
1. a hgr.s file with just what you suggest;
2. the pragmas in `apple2_hgr_init_graphics.c` as you suggest;
3. compile as follows:
```
apple2_hgr:	 
	$(CC65_PATH)$(MYCC65) $(INCLUDE_OPTS) -t apple2 \
	--config $(CFG_PATH)/cc65/apple2-hgr.cfg \
...
	$(CROSS_LIB_PATH)/sound/generic/bit_bang_sounds.c \
	$(CROSS_LIB_PATH)/sleep/sleep_macros.c \
	$(CROSS_LIB_PATH)/display/init_graphics/cc65/apple2/hgr.s \
	$(FULL_FILES) \
	$(CROSS_LIB_PATH)/display/init_graphics/cc65/apple2/apple2_hgr_init_graphics.c \
    --code-name LOWCODE \
	$(SOURCE_PATH)/end_screen.c \
	-o $(BUILD_PATH)/apple2.bin
```

P.S.: I cannot use -O in CC65 because I am using some generic C code for the big bang sounds that CC65 optimizer breaks because of the lack of volatile support.

**Reply by greg-king5** [reply]
> If I try to follow your suggestion, I get:
> ```
> cross_lib/display/init_graphics/cc65/apple2/apple2_hgr_init_graphics.c(129): Warning: Variable 'bitmap
> ' is defined but never used
> ld65: Warning: cross_lib/cfg//cc65/apple2-hgr.cfg(27): Segment 'CODE' start address is too low in 'MAI
> N' by 8192 bytes
> ```

I forgot about that optimization (sometimes, the compiler won't output _static_ global objects that aren't referenced).  Therefore, remove the `static` keyword from bitmap's definition.

> ```
> ld65: Warning: cross_lib/cfg//cc65/apple2-hgr.cfg(17): Segment 'CODE' overflows memory area 'MAIN' by
> 2274 bytes
> ld65: Error: Cannot generate most of the files due to memory area overflows
> ```

If you still get an "overflow" message, then you should try to move more things into the `LOWCODE` segment.  If those moves still don't solve it, then you will need to reduce the features of your games.

> What I have done is to have:
> 
>     1. a hgr.s file with just what you suggest;
> 
>     2. the pragmas in `apple2_hgr_init_graphics.c` as you suggest;

Step number 2 _replaces_ step number 1.  Therefore, delete the "hgr.s" Assembly file.

**Comment by Fabrizio-Caruso** [comment]
@greg-king5 Following my previous set-up (not following your suggestions). I can get my game to work but If I define one more array in `apple2_hgr_init_graphics.c`, the program will just crash at start-up no matter whether it is put in LOWCODE or CODE.
I can produce a .map file but I am not able to spot the problem.
Some code is overwritten somehow:

I have attached the .map file here.
[apple2.zip](https://github.com/cc65/cc65/files/6534027/apple2.zip)

**Comment by Fabrizio-Caruso** [comment]
I think I found the problem: the binary is too big.
